opamp/observiq/agent_config_manager.go

```diff
@@ -0,0 +1,215 @@
+			ContentType: opamp.DetermineContentType(managedConfig.ConfigPath),
+		}
+	}
+
+	return &protobufs.EffectiveConfig{
+		ConfigMap: &protobufs.AgentConfigMap{
+			ConfigMap: contentMap,
+		},
+	}, nil
+}
+
+// ApplyConfigChanges compares the remoteConfig to the existing and applies changes
+func (a *AgentConfigManager) ApplyConfigChanges(remoteConfig *protobufs.AgentRemoteConfig) (effectiveConfig *protobufs.EffectiveConfig, changed bool, returnErr error) {
+	// Always compute effective config at the end. This ensures we always have the most up to date copy when we respond to the server
+	defer func() {
+		var err error
+		effectiveConfig, err = a.ComposeEffectiveConfig()
+		if err != nil {
+			a.logger.Error("Failed to compute effective config while applying config changes", zap.Error(returnErr))
+		}
+	}()
+
+	remoteConfigMap := remoteConfig.GetConfig().GetConfigMap()
+
+	// No remote config Map
+	if remoteConfigMap == nil {
+		return
+	}
+
+	// loop through all remote configs and compare then with existing configs
+	for configName, remoteContents := range remoteConfigMap {
+		// For security check the log file we want is acceptable
+		if _, ok := acceptableConfigs[configName]; !ok {
+			a.logger.Warn("Not supported config received skipping", zap.String("config", configName))
+			continue
+		}
+
+		managedConfig, ok := a.configMap[configName]
+		// This check is impossible to hit now but will be a use case we'll have in the future so leaving this tested code in for now.
+		if !ok {
+			// We don't current track this config file we should add it
+			if err := a.trackNewConfig(configName, remoteContents.GetBody()); err != nil {
+				returnErr = err
+				return
+			}
+			changed = true
+			continue
+		}
+
+		// Update the config file
+		configChanged, err := a.updateExistingConfig(configName, managedConfig, remoteContents.GetBody())
+		if err != nil {
+			returnErr = err
+			return
+		}
+
+		changed = changed || configChanged
+	}
+
+	return
+}
+
+func (a *AgentConfigManager) updateExistingConfig(configName string, managedConfig *opamp.ManagedConfig, newContents []byte) (changed bool, err error) {
+	remoteHash := opamp.ComputeHash(newContents)
+
+	// Nothing to update
+	if bytes.Equal(managedConfig.GetCurrentConfigHash(), remoteHash) {
+		// Verify disk matches memory and remote. This will ensure if any changes were made to the config on disk
+		// the will be overwritten by what is expected.
+		return verifyDiskContents(managedConfig.ConfigPath, managedConfig.GetCurrentConfigHash(), newContents)
+	}
+
+	a.logger.Info("Applying changes to config file", zap.String("config", configName))
+	changed, err = managedConfig.Reload(newContents)
+	if err != nil {
+		err = fmt.Errorf("failed to reload config: %s: %w", configName, err)
+		return
+	}
+
+	// If the config changed recompute the hash for it
+	if changed {
+		err = managedConfig.ComputeConfigHash()
+		if err != nil {
+			err = fmt.Errorf("failed hash compute for config %s: %w", configName, err)
```

Comment:
this seems like a pretty bad case. like we've updated our config info in memory, but something is jacked where we couldn't calculate the hash (missing file on disk?).

actually thinking this through maybe it's not a problem. cause if there was an old hash that didn't match the current config (because computing the new hash failed), then the next time iris requests aggregated data we won't give it the missing file data and it'll know there's a problem and send another config maybe which hopefully we succeed next time.

but if the error was purely with calculating the hash and it was temporary (i don't know why this would be the case), that might be more problematic. like if the new file was written, but the hash wasn't computed and the stored one is old. could this get funky? do we need a comment here or a backlog tech debt to look into this?

Reply:
So the hash function fails if we fail to open the file to read. Which means we don't recompute the hash. It does put things in a weird state where likely you are running with the correct config and have the correct one on disk at this point. So on the next apply remote config it'll show a mismatch and apply the config again.

Due to the amount of file ops the entire reload logic entails there are several errors points. It's more likely though to have all of them fail rather than one randomly unless the system is running out of file descriptors. I think this is an acceptable risk and the case where we're wrong seems like it would be self correcting eventually.

Reply:
@cpheps yeah, i think no matter what this ends up self correcting. i'll resolve